### ISSUE TITLE
Authorization via Azure AD / RBAC

### ISSUE THREAD

**Issue by ItalyPaleAle** [Azure/azure-storage-blob-go#160]
Using the SDK version v38.2.0, I cannot find a way to authorize the storage SDK using Azure AD.

I'd like to use Azure SDK for Go (not storage) and authorize access to other services using a service principal (via env vars `AZURE_TENANT_ID`, `AZURE_CLIENT_ID`, `AZURE_CLIENT_SECRET`). This doesn't seem to be possible using the Blob Storage SDK for Go, however.

I have found in the code the `NewTokenCredential` class which seems to be related, but I can't find any information on how to use it (if it's actually related)

**Comment by lukaszbudnik** [comment]
Agree, it would be nice if there was an option to use Authorizer just like the rest of the AWS SDK: https://github.com/Azure/azure-sdk-for-go#authentication

**Comment by tombuildsstuff** [comment]
@ItalyPaleAle @lukaszbudnik FWIW this SDK uses a different underlying transport mechanism to the regular Azure SDK for Go which is why these aren't compatible with the [the SharedKey](https://github.com/Azure/go-autorest/blob/master/autorest/authorization_storage.go#L78) and the [Bearer (AzureAD auth) Token](https://github.com/Azure/go-autorest/blob/master/autorest/authorization.go#L107) Authorizer found in that package. I appreciate it's a little cheeky to cross-link here - but I think it's relevant to the problem at hand:

 For various reasons we ended up implementing an alternate SDK for the Azure Storage API's ([Giovanni](https://github.com/tombuildsstuff/giovanni)) which uses [the same underlying transport layer](https://github.com/Azure/go-autorest) as the main [Azure SDK for Go](https://github.com/Azure/azure-sdk-for-go) - including [the SharedKey authorizer](https://github.com/Azure/go-autorest/blob/master/autorest/authorization_storage.go#L78) and the [Bearer (AzureAD auth) Token Authorizer](https://github.com/Azure/go-autorest/blob/master/autorest/authorization.go#L107) both of which work with the Storage API's and may be what you're looking for here?

Hope that helps!

**Comment by ItalyPaleAle** [comment]
Hey Tom, I tried your SDK but it didn't work for me. Aside from some issues getting started (the documentation isn't complete, and there was an issue with the sample code showing authentication), it was lacking some features I absolutely needed, such as the ability to download a file into a stream (io.Reader rather than a byte slice).

I was able to figure out how to use an Azure AD token with the official Storage SDK. As soon as I clean up the code I'll post a Gist and add a link here

**Comment by ItalyPaleAle** [comment]
Here's the Gist: https://gist.github.com/ItalyPaleAle/ec6498bfa81a96f9ca27a2da6f60a770

Hopefully it can help others

**Comment by tombuildsstuff** [comment]
@ItalyPaleAle cool, glad to hear you got this one working 👍

**Comment by lukaszbudnik** [comment]
good to know, quick question - does it work with MSI?

**Comment by guitmz** [comment]
@lukaszbudnik this works for MSI, just spent 3 freaking days trying to figure this out with a coworker. Azure is great!

```
func getAzureStorageCredentials() (azblob.Credential, error) {
	azureServicePrincipalToken, err := auth.NewMSIConfig().ServicePrincipalToken()
	if err != nil {
		log.Fatal(err)
	}

	// Get a new token
	err = azureServicePrincipalToken.Refresh()
	if err != nil {
		log.Fatal(err)
	}
	token := azureServicePrincipalToken.Token()

	// Credential object
	credential := azblob.NewTokenCredential(token.AccessToken, nil)
	return credential, nil
}
```

**Comment by lukaszbudnik** [comment]
@guitmz that actually worked - thanks a million!
